include/swift/Demangling/TypeDecoder.h

```diff
@@ -1379,28 +1380,39 @@ class TypeDecoder {
         node = node->getFirstChild();
         hasParamFlags = true;
       };
-      switch (node->getKind()) {
-      case NodeKind::InOut:
-        setOwnership(ValueOwnership::InOut);
-        break;
 
-      case NodeKind::Shared:
-        setOwnership(ValueOwnership::Shared);
-        break;
+      bool recurse = true;
+      while (recurse) {
+        switch (node->getKind()) {
+        case NodeKind::InOut:
+          setOwnership(ValueOwnership::InOut);
```

Comment:
This will result in an infinite loop. You should set the ``recurse`` flag only to true for cases where you also change ``node``. E.g.
```
bool recurse = false;
do {
  recurse = false;
  switch (node->getKind()) {
    ...
    case NodeKind::NoDerivative:
      node = node->getFirstChild();
      recurse = true;
      break;
    ...
} while (recurse);
```
Also, can you add a test for this?

Reply:
The current implementation won't case infinite loops, because `setOwnership` always changes `node` to its child.

Reply:
ah, I didn't see that ``setOwnership`` changes node.

Reply:
IMO, this is really subtle. It would be much simpler to just manually inline the few lines of ``setOwnership``.

Reply:
Sounds good. I've inlined them.